cmd/bucket-metadata-sys.go

```diff
@@ -420,6 +420,35 @@ func (sys *BucketMetadataSys) GetConfig(bucket string) (BucketMetadata, error) {
 	return meta, nil
 }
 
+// IsTierInUse returns true if there exists an ILM transition rule storage-class
+// referring to tierName and false otherwise. N B this check is potentially
+// racy. It doesn't protect the caller from the effects of a concurrent
+// put-bucket-lifecycle/delete-bucket-lifecycle operation. It is only a
+// best-effort protection to prevent users from deleting a remote tier which is
+// in active use.
+func (sys *BucketMetadataSys) IsTierInUse(tierName string) bool {
+	sys.RLock()
+	defer sys.RUnlock()
+
+	for _, bucketMeta := range sys.metadataMap {
+		// skip buckets with no ILM rules
+		if bucketMeta.lifecycleConfig == nil {
+			continue
+		}
+		for _, rule := range bucketMeta.lifecycleConfig.Rules {
+			// skip disabled ILM rules
+			if rule.Status == lifecycle.Disabled {
+				continue
+			}
+			if rule.Transition.StorageClass == tierName {
+				return true
+			}
+		}
```

Comment:
this by itself is not sufficient - some objects could have transitioned to this tier and the lifecycle rule could have been disabled or removed entirely. This will cause objects transitioned previously from being accessible if the tier config is deleted. Seems to me this has to be cached during scanner run to see if tier is in use.

Reply:
yes, that's correct. I shall include additional validation to prevent the case you describe above. The idea is to list objects in the remote tier, allow tier to be deleted only if there was no objects found. This way we ensure that the remote-tier is not used by active ILM rules and that there are no previously transitioned objects. Does this make sense?